datocms/gatsby-source-datocms
# Content sync support

This add functionality for the source plugin to create node manifests. These manifests are used in tandem with a new feature we are rolling out called Content Sync. This will make it so we can link content in a CMS that a user wants to preview to the exact page and build that the user wants to preview and route the user to that content as soon as it's ready. Previously if the page was not created yet it would 404, leading to a confusing experience for the CMS user.

These source plugin changes are going out together with changes to https://github.com/gatsby-inc/datocms-plugin-gatsby-cloud to improve Gatsby’s Preview service on DatoCMS. The user is now redirected to a loading page on Gatsby Cloud when they press “Open Preview”. The loading page waits for the associated build to complete and then redirects the user to the correct page once the updated content is available. The loading page also handles build errors and will let the user know if the build they are waiting for fails. 
To support these changes, we also updated the DatoCMS Source Plugin for Gatsby. The updated plugin code and test website it located here: https://github.com/gatsby-inc/datocms-content-sync . The source plugin changes are in the plugins directory of this Gatsby site since we're not able to publish datocms canaries.

To test out this functionality: 
- Clone the https://github.com/gatsby-inc/datocms-content-sync repo. (or create a canary for the plugin and use any datocms site)
- Create a demo project using Gatsby Blog Demo Starter (linked from repo ReadMe) and host on Gatsby Cloud. Make sure to click on “Enable CMS Preview Builds” under Preview in the settings. 
- Add the new preview feature as a private plugin on DatoCMS. 
- The app is hosted here https://datocms-app-plugin.netlify.app/ and this can be used as the entry point in the plugin settings. 
- Add the following parameter definitions to the `”global”` array:
```
  {
      "id": "instanceUrl",
      "label": "Preview Instance URL",
      "type": "string",
      "hint": "The url of a Gatsby Cloud Preview instance"
    },
  	{
      "id": "contentSyncUrl",
      "label": "Content Sync URL",
      "type": "string",
      "hint": "https://gatsbyjs.com/content-sync/#{site_id}"
    }
```
 NOTE:
 - Check the CMS preview tab for the instance URL value.
 - The Content Sync URL will be https://gatsbyjs.com/content-sync/#{site_id} . The value for the site id can be copied from end of the web hook URLs under site settings on Gatsby Cloud.
 - Also copy the preview and build webbook URLs from the general site settings on Gatsby Cloud to create web hooks on DatoCMS.
